### PR TITLE
Fix state mismatch

### LINKED ISSUE #2083: Mainnet state mismatch
**Describe the bug**

It appeared on 17.10.2022 on the test machine run after commit ef9f9a4. It's a potential regression.

**To Reproduce**
1. checkout commit ef9f9a4
2. Run `forest` on mainnet for some time
3. Hopefully see the error

**Log output**

<details>
  <summary>Log Output</summary>
  
  ```Paste log output here
 2022-10-17T17:25:24.456Z INFO  forest::daemon > Starting Forest daemon, version 0.4.1+git.ef9f9a4
 2022-10-17T17:25:24.456Z INFO  forest_libp2p::service > Recovered libp2p keypair from "/opt/forest/libp2p/keypair"
 2022-10-17T17:25:24.456Z INFO  forest::daemon         > PeerId: 12D3KooWQwtGB2d5Z5SbiLXD8KKWWoSZax9jWJa2LtXgg2xKeEm9
 2022-10-17T17:25:24.457Z WARN  forest::daemon         > Warning: Keystore encryption disabled!
 2022-10-17T17:25:24.457Z INFO  forest::daemon         > Prometheus server started at 0.0.0.0:6116
 2022-10-17T17:25:24.457Z INFO  forest::daemon         > Admin token: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.BM22GMXwD6Se98Q-g6Vj1iFKB72RD5X8qlmWc_6urf8
 2022-10-17T17:25:25.360Z INFO  forest_genesis         > Initialized genesis: BlockHeader: Cid(bafy2bzacecnamqgqmifpluoeldx7zzglxcljo6oja4vrmtj7432rphldpdmm2)
 2022-10-17T17:25:25.360Z INFO  forest::daemon         > Using network :: testnetnet
 2022-10-17T17:25:25.363Z INFO  forest_chain_sync::chain_muxer > Evaluating network head...
 2022-10-17T17:25:25.363Z INFO  forest::daemon                 > JSON-RPC endpoint started at 127.0.0.1:1234
 2022-10-17T17:25:25.363Z INFO  forest_rpc                     > Ready for RPC connections
 2022-10-17T17:25:26.686Z INFO  forest_chain_sync::chain_muxer > Local node is behind the network, starting BOOTSTRAP from LOCAL_HEAD = 2256489 -> NETWORK_HEAD = 2257370
 2022-10-17T17:25:27.561Z INFO  forest_chain_sync::tipset_syncer > Validating tipset: EPOCH = 2256489
 2022-10-17T17:25:27.695Z INFO  forest_chain_sync::tipset_syncer > Validating tipset: EPOCH = 2256490
StateDiff:
  Expected: bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72
  Root: bafy2bzacebp76qjdtbcaek2ko66byihpesvd22lmtzd27gcheqdthu2az2sum
StateDiff:
  Expected: bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72
  Root: bafy2bzacebp76qjdtbcaek2ko66byihpesvd22lmtzd27gcheqdthu2az2sum
StateDiff:
  Expected: bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72
  Root: bafy2bzacebp76qjdtbcaek2ko66byihpesvd22lmtzd27gcheqdthu2az2sum
StateDiff:
  Expected: bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72
  Root: bafy2bzacebp76qjdtbcaek2ko66byihpesvd22lmtzd27gcheqdthu2az2sum
StateDiff:
  Expected: bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72
  Root: bafy2bzacebp76qjdtbcaek2ko66byihpesvd22lmtzd27gcheqdthu2az2sum
Downloading headers 0 / 881 [-----------------------------------] 0.00 % 0.00/s 
Downloading headers 622 / 881 [==================>-------] 70.60 % 1240.76/s 0s 
Downloading headers 881 / 881 [============================] 100.00 % 1106.59/s 
Downloading headers 881 / 881 [============================] 100.00 % 1106.55/s Could not resolve actor states: failed to find state tree bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72
Using default resolution:
Could not resolve actor states: failed to find state tree bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72
Using default resolution:
Could not resolve actor states: failed to find state tree bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72
Using default resolution:
Could not resolve actor states: failed to find state tree bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72
Using default resolution:
Could not resolve actor states: failed to find state tree bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72
Using default resolution:
Failed to print state-diff: Cid does not exist in blockstore
Failed to print state-diff: Cid does not exist in blockstore
Failed to print state-diff: Cid does not exist in blockstore
StateDiff:
  Expected: bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72
  Root: bafy2bzacebp76qjdtbcaek2ko66byihpesvd22lmtzd27gcheqdthu2az2sum
Failed to print state-diff: Cid does not exist in blockstore
Failed to print state-diff: Cid does not exist in blockstore
StateDiff:
  Expected: bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72
  Root: bafy2bzacebp76qjdtbcaek2ko66byihpesvd22lmtzd27gcheqdthu2az2sum
StateDiff:
  Expected: bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72
Could not resolve actor states: failed to find state tree bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72
Using default resolution:
Could not resolve actor states: failed to find state tree bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72
Using default resolution:
Could not resolve actor states: failed to find state tree bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72
Using default resolution:
  Root: bafy2bzacebp76qjdtbcaek2ko66byihpesvd22lmtzd27gcheqdthu2az2sum
StateDiff:
  Expected: bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72
  Root: bafy2bzacebp76qjdtbcaek2ko66byihpesvd22lmtzd27gcheqdthu2az2sum
Failed to print state-diff: Cid does not exist in blockstore
Failed to print state-diff: Cid does not exist in blockstore
Failed to print state-diff: Cid does not exist in blockstore
Could not resolve actor states: failed to find state tree bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72
Using default resolution:
Failed to print state-diff: Cid does not exist in blockstore
 2022-10-17T17:26:09.279Z WARN  forest_chain_sync::tipset_syncer > Validating block [CID = bafy2bzacecgtzq7gzsyqfmcia2enuxwzzjpfbmcmj3274swsjbaxeszbu3ufy] in EPOCH = 2256490 failed: Validation error: Validation error: Parent state root did not match computed state: bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72 (header), bafy2bzacebp76qjdtbcaek2ko66byihpesvd22lmtzd27gcheqdthu2az2sum (computed)
 2022-10-17T17:26:09.279Z ERROR forest_chain_sync::tipset_syncer > Sync messages check state failed for tipset range
 2022-10-17T17:26:09.279Z ERROR forest_chain_sync::chain_muxer   > Bootstrapping failed, re-evaluating the network head to retry the bootstrap. Error = TipsetRangeSyncer(Validation("Validation error: Parent state root did not match computed state: bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72 (header), bafy2bzacebp76qjdtbcaek2ko66byihpesvd22lmtzd27gcheqdthu2az2sum (computed)"))
 2022-10-17T17:26:09.279Z INFO  forest_chain_sync::chain_muxer   > Evaluating network head...
 2022-10-17T17:26:15.923Z INFO  forest_chain_sync::chain_muxer   > Local node is behind the network, starting BOOTSTRAP from LOCAL_HEAD = 2256489 -> NETWORK_HEAD = 2257372
 2022-10-17T17:26:15.929Z ERROR forest_chain_sync::chain_muxer   > Bootstrapping failed, re-evaluating the network head to retry the bootstrap. Error = TipsetRangeSyncer(TipsetRangeWithBadBlock(Cid(bafy2bzacecgtzq7gzsyqfmcia2enuxwzzjpfbmcmj3274swsjbaxeszbu3ufy), "Validation error: Validation error: Parent state root did not match computed state: bafy2bzacec2zcy52ypkq2pnupw2tqan2ioev2maqrronq76otfyjyjdxara72 (header), bafy2bzacebp76qjdtbcaek2ko66byihpesvd22lmtzd27gcheqdthu2az2sum (computed)"))
 2022-10-17T17:26:15.929Z INFO  forest_chain_sync::chain_muxer   > Evaluating network head...
 2022-10-17T17:26:33.053Z INFO  forest_chain_sync::chain_muxer   > Local node is behind the network, starting BOOTSTRAP from LOCAL_HEAD = 2256489 -> NETWORK_HEAD = 2257373
  ```
</details> 

**Expected behaviour**

No state mismatches

**Screenshots**

**Environment (please complete the following information):**
 - OS: Fedora 36
 - Rust version(e.g. `rustc --version)  https://github.com/ChainSafe/forest/blob/main/rust-toolchain.toml
 - Branch/commit ef9f9a4 (`main`)

**Other information and links**

### PR DESCRIPTION
**Pull request by elmattic**
**Summary of changes**
Changes introduced in this pull request:
- Refactor `verify_consensus_fault` to make it look easier to compare with Lotus
- Don't return an error on `verify_block_signature` failures

What's still needed:

- [x] Check height 2,226,489
- [x] Check height 2,293,819
- [x] Sync up to HEAD
- [ ] Tests?

**Reference issue to close (if applicable)**

Closes #2083

**Other information and links**

### CODE REVIEW

**Comment by lemmih** [comment]
Testing epoch 2293819 now.

**Comment by lemmih** [comment]
> Testing epoch 2293819 now.

Works for me. Yay!

**Comment by elmattic** [comment]
> > Testing epoch 2293819 now.
> 
> Works for me. Yay!

Great! Did you validate the whole snapshot?

**Comment by lemmih** [comment]
> > > Testing epoch 2293819 now.
> > 
> > 
> > Works for me. Yay!
> 
> Great! Did you validate the whole snapshot?

No, just a few dozen tipsets around the problematic epoch.

**Comment by lemmih** [comment]
I'll sync to HEAD and keep it running for a few days.

File: vm/interpreter/src/fvm.rs [hunk 1]
```diff
@@ -164,62 +165,67 @@ impl<DB: Blockstore> Consensus for ForestExterns<DB> {
...
-                })
+                fault_type = Some(ConsensusFaultType::ParentGrinding);
             }
         };
 
         // (3) return if no consensus fault
-        if cf.is_none() {
-            return Ok((cf, total_gas));
+        if fault_type.is_none() {
+            return Ok((ret, total_gas));
```
Comment (lemmih) [inline vm/interpreter/src/fvm.rs:202]: `ret` is always `None` at this point, right?

File: vm/interpreter/src/fvm.rs [hunk 2]
```diff
@@ -164,62 +165,67 @@ impl<DB: Blockstore> Consensus for ForestExterns<DB> {
...
+        if let Ok(gas_used) = self.verify_block_signature(&bh_1) {
+            total_gas += gas_used;
+            if let Ok(gas_used) = self.verify_block_signature(&bh_2) {
+                total_gas += gas_used;
+                ret = Some(ConsensusFault {
+                    target: *bh_1.miner_address(),
+                    epoch: bh_2.epoch(),
+                    // Unwrapping is safe here, see (3)
+                    fault_type: fault_type.unwrap(),
+                });
```
Comment (lemmih) [inline vm/interpreter/src/fvm.rs:219]: This seems to be the only place where `ret` is given a value. You wrote it like this to match the golang code?

File: vm/interpreter/src/fvm.rs (hunk 1 above)
Comment (elmattic) [reply vm/interpreter/src/fvm.rs:202]: Yes.

**Comment by lemmih** [comment]
I think you found the bug. Let's clean up the code and get it merged.

**Comment by elmattic** [comment]
> I think you found the bug. Let's clean up the code and get it merged.

What do you suggest for cleaning it up?

**Comment by LesnyRumcajs** [comment]
@elmattic could you point us to a relevant piece of logic in Lotus?

**Comment by elmattic** [comment]
> @elmattic could you point us to a relevant piece of logic in Lotus?

Of course.

https://github.com/filecoin-project/lotus/blob/master/chain/vm/fvm.go#L111-L225

**Comment by lemmih** [comment]
> > @elmattic could you point us to a relevant piece of logic in Lotus?
> 
> Of course.
> 
> https://github.com/filecoin-project/lotus/blob/master/chain/vm/fvm.go#L111-L225

Let's add that link to the code.

**Comment by lemmih** [comment]
I wish we had a way to write a regression test for this. Might not be feasible with the way the `ForestExtern` is currently structured, though.

File: vm/interpreter/src/fvm.rs [hunk 3]
```diff
@@ -198,28 +191,39 @@ impl<DB: Blockstore> Consensus for ForestExterns<DB> {
...
-        Ok((cf, total_gas))
+        if let Ok(gas_used) = self.verify_block_signature(&bh_1) {
+            total_gas += gas_used;
+            if let Ok(gas_used) = self.verify_block_signature(&bh_2) {
+                total_gas += gas_used;
+                let ret = Some(ConsensusFault {
+                    target: *bh_1.miner_address(),
+                    epoch: bh_2.epoch(),
+                    // Unwrapping is safe here, see (3)
+                    fault_type: fault_type.unwrap(),
```
Comment (lemmih) [inline vm/interpreter/src/fvm.rs:216]: We could use a match instead of `fault_type.is_none()` to get rid of this `unwarp`.

**Review by lemmih — APPROVED** [review]

**Comment by elmattic** [comment]
> I wish we had a way to write a regression test for this. Might not be feasible with the way the `ForestExtern` is currently structured, though.

I think so too, should I open a new issue to track this?

**Comment by lemmih** [comment]
> > I wish we had a way to write a regression test for this. Might not be feasible with the way the `ForestExtern` is currently structured, though.
> 
> I think so too, should I open a new issue to track this?

Yeah, let's open a tracking issue.

**Review by LesnyRumcajs — APPROVED** [review]

**Pull request merged**